asyncnsq/tcp/reader.py

```diff
@@ -173,8 +193,62 @@ async def _redistribute(self):
     async def _lookupd(self):
         host, port = random.choice(self._lookupd_http_addresses)
         await self._poll_lookupd(host, port)
-
-
-    def close(self):
-        for conn in self._connections:
-            conn.close()
+    
+    async def unsubscribe(self):
```

Comment:
why and when a reader needs a unsubscribe method?  what's it situation?

Reply:
It is necessary for our application to subscribe different topics, and sometimes we need to cancel the consuming due to some external events.

For example, our application is registered in zookeeper or etcd. Once we detected some rescheduling events, we need to cancel our current messages or jobs, renqueue all the unfinished work. In the meantime, some other instences will take over these jobs, and hoping to continue it smothly. 


suppose we have 2 functions

```py
async def message_consuming_loop():
    async for message in reader.messages():
        # got a message, trigger a few events, fin or req later
        pass
    # jump out from the loop, and handle rest of the process
    for msg in msgs_not_really_finished:
         # maybe not finished in batch processing
         msg.req(10) # defer it, try it later
    await reader.cancel() # clean up all the connections, clean up in-flight message safely
```

```py
async def interrupt_the_consuming():
      ''' we noticed some external event, and we are supposed to let 
          message_consuming_loop jump out
      '''
      reader. unsubscribe()
```

once the interrupt_the_consuming was called, `for message in reader.messages()` loop will got jumped out, and we can gracefully hand the unfinished messages, and close all the connections clearly.

Reply:
the `jump out from the loop, and handle rest of the process` stage will be triggered after `reader.unsubscribe()` called.
At this moment, all the connections are keeping alive, but do not have any new messages now.
We may still have some unfinished jobs, we can req(0) them, and then call `await cancel()` or `close` to close all the connections

Reply:
ok, i see . you hanlde many topics in one reader, and you want to manipulate some of it.

would you mind add more unitest for this feature?

Reply:
> would you mind add more unitest for this feature?

👌

Reply:
@aohan237 I had a few test cases in my project, I was supposed to append them to the `tests/test_reader_and_writer.py` but failed. 

```
$ python runtests.py 
Traceback (most recent call last):
  File "runtests.py", line 291, in <module>
    runtests()
  File "runtests.py", line 250, in runtests
    self.logger = logging.getself.logger()
AttributeError: module 'logging' has no attribute 'getself'
$ python tests/test_reader_and_writer.py 
Traceback (most recent call last):
  File "tests/test_reader_and_writer.py", line 2, in <module>
    from ._testutils import run_until_complete, BaseTest
ImportError: attempted relative import with no known parent package
```

It could be another pr?